src/operator/join.rs

```diff
@@ -311,4 +377,99 @@ mod test {
+                finite_map! { (2, 3) => 1, (1, 3) => 1 },
+                finite_map! { (1, 3) => 1, (2, 3) => 1, (2, 4) => 1, (2, 1) => 1, (4, 1) => 1, (4, 3) => 1 },
+                finite_map! { (1, 1) => 1, (2, 2) => 1, (3, 3) => 1, (4, 4) => 1,
+                              (1, 2) => 1, (1, 3) => 1, (1, 4) => 1,
+                              (2, 1) => 1, (2, 3) => 1, (2, 4) => 1,
+                              (3, 1) => 1, (3, 2) => 1, (3, 4) => 1,
+                              (4, 1) => 1, (4, 2) => 1, (4, 3) => 1 },
+            ]
+            .into_iter();
+
+            let edges: Stream<_, FiniteHashMap<(usize, usize), isize>> =
+                circuit
+                    .add_source(Generator::new(move || edges.next().unwrap()));
+
+            let paths = circuit.iterate_with_conditions(|child| {
+                // ```text
+                //                      distinct_incremental_nested
+                //               ┌───┐          ┌───┐
+                // edges         │   │          │   │  paths
+                // ────┬────────►│ + ├──────────┤   ├────────┬───►
+                //     │         │   │          │   │        │
+                //     │         └───┘          └───┘        │
+                //     │           ▲                         │
+                //     │           │                         │
+                //     │         ┌─┴─┐                       │
+                //     │         │   │                       │
+                //     └────────►│ X │ ◄─────────────────────┘
+                //               │   │
+                //               └───┘
+                //      join_incremental_nested
+                // ```
+                let edges = edges.delta0(child);
+                let paths_delayed = DelayedFeedback::new(child);
+
+                let paths_inverted: Stream<_, FiniteHashMap<(usize, usize), isize>> = paths_delayed
+                    .stream()
+                    .apply(|paths: &FiniteHashMap<(usize, usize), isize>| paths.into_iter().map(|((x,y), w)| ((*y, *x), *w)).collect());
+
+                let paths_inverted_indexed: Stream<_, FiniteHashMap<usize, FiniteHashMap<usize, isize>>> = paths_inverted.index();
+                let edges_indexed: Stream<_, FiniteHashMap<usize, FiniteHashMap<usize, isize>>> = edges.index();
+
+                let paths = edges.plus(&paths_inverted_indexed.join_incremental_nested(&edges_indexed, |_via, from, to| (*from, *to)))
+                    // Non-incremental implementation of distinct_nested_incremental.
+                    .integrate()
+                    .integrate_nested()
+                    .apply(|i| i.distinct())
+                    .differentiate()
+                    .differentiate_nested();
+                paths_delayed.connect(&paths);
+                let output = paths.integrate();
+                Ok((
+                    vec![
+                        paths.condition(|delta| delta.support_size() == 0),
```

Comment:
why are both of these needed?
Isn't support_size() == 0 as efficient as is_zero()? I find is_zero clearer and cleaner.

Reply:
> why are both of these needed?

This is subtle stuff and we should discuss it, ideally in person.  It is related to the semantics of z-1 with finite prefixes.

> Isn't support_size() == 0 as efficient as is_zero()? I find is_zero clearer and cleaner.

yep, thanks!